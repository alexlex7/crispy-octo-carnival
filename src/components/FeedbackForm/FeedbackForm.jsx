import React, { useState } from 'react';
import styles from './FeedbackForm.module.css';
import PhoneInput from 'react-phone-input-2';
import axios from 'axios';
import StarRating from '../StarRating/StarRating';
import PropTypes from 'prop-types';

const initialState = {
  firstName: '',
  lastName: '',
  phone: '',
  email: '',
  feedback: '',
  department: 'value1',
  permission: false,
  rating: 0,
};

const initialErrors = {
  firstName: { valid: true, message: '' },
  lastName: { valid: true, message: '' },
  phone: { valid: true, message: '' },
  email: { valid: true, message: '' },
  feedback: { valid: true, message: '' },
};

export default function FeedbackForm({ onSubmitForm }) {
  const [formData, setFormData] = useState(initialState);
  const [validationErrors, setValidationErrors] = useState(initialErrors);

  const handlePhoneChange = (value) => {
    setFormData({ ...formData, phone: value });
  };

  const handleChangeInputs = (e) => {
    const { name } = e.target;
    switch (name) {
      case 'firstName':
        setFormData({ ...formData, firstName: e.target.value });
        break;
      case 'lastName':
        setFormData({ ...formData, lastName: e.target.value });
        break;
      case 'email':
        setFormData({ ...formData, email: e.target.value });
        break;
      case 'feedback':
        setFormData({ ...formData, feedback: e.target.value });
        break;
      case 'department':
        setFormData({ ...formData, department: e.target.value });
        break;
      case 'permission':
        setFormData({ ...formData, permission: e.target.checked });
        break;

      default:
        break;
    }
  };

  const validate = (name, value) => {
    switch (name) {
      case 'firstName':
        if (value.length < 2) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'First name must be 2 char. min.',
            },
          });
        } else {
          setValidationErrors({
            ...validationErrors,
            [name]: { valid: true, message: '' },
          });
        }
        break;
      case 'lastName':
        if (value.length < 2) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'Last name must be 2 char. min.',
            },
          });
        } else {
          setValidationErrors({
            ...validationErrors,
            [name]: { valid: true, message: '' },
          });
        }
        break;
      case 'email':
        if (!value.match(/@/)) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'Email must be valid.',
            },
          });
        } else {
          setValidationErrors({
            ...validationErrors,
            [name]: { valid: true, message: '' },
          });
        }
        break;
      case 'feedback':
        if (value.length < 100) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'Feedback must be 100 char. min.',
            },
          });
        } else if (value.length > 150) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'Feedback must be 150 char. max.',
            },
          });
        } else {
          setValidationErrors({
            ...validationErrors,
            [name]: { valid: true, message: '' },
          });
        }
        break;
      case 'phone':
        if (value.length < 19) {
          setValidationErrors({
            ...validationErrors,
            [name]: {
              valid: false,
              message: 'Phone must be valid',
            },
          });
        } else {
          setValidationErrors({
            ...validationErrors,
            [name]: { valid: true, message: '' },
          });
        }
        break;
      default:
        break;
    }
  };

  const validationHandler = (e) => {
    const { name, value } = e.target;
    validate(name, value);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    onSubmitForm(formData);
    // try {
    //   await axios({
    //     method: 'post',
    //     url: 'http://localhost:8080/feedback',
    //     data: {
    //       ...formData,
    //     },
    //   });
    //   await onSubmitForm(formData);
    //   toggleModal();
    // } catch (error) {
    //   console.log(error);
    // }
    setFormData(initialState);
  };

  const handleSetRating = (rating) => {
    const countRaiting = Math.round(((rating / 200).toPrecision(2) * 100) / 10);
    setFormData({ ...formData, rating: (5 / 100) * (countRaiting * 10) });
  };

  return (
    <form className={styles.form} onSubmit={handleSubmit}>
      <label className={styles.label}>
        First name
        <input
          className={styles.input}
          type="text"
          name="firstName"
          value={formData.firstName}
          onChange={handleChangeInputs}
          required
          minLength="1"
          placeholder="enter your first name"
          onBlur={validationHandler}
        />
        {!validationErrors.firstName.valid ? (
          <span className={styles.error}>
            {validationErrors.firstName.message}
          </span>
        ) : null}
      </label>
      <label className={styles.label}>
        Last name
        <input
          className={styles.input}
          type="text"
          name="lastName"
          value={formData.lastName}
          onChange={handleChangeInputs}
          required
          minLength="2"
          placeholder="enter your last name"
          onBlur={validationHandler}
        />
        {!validationErrors.lastName.valid ? (
          <span className={styles.error}>
            {validationErrors.lastName.message}
          </span>
        ) : null}
      </label>
      <label className={styles.label}>
        Phone
        <PhoneInput
          inputProps={{ name: 'phone' }}
          buttonClass={styles.dropdownBtn}
          inputClass={styles.phoneInput}
          inputStyle={{ width: '100%' }}
          value={formData.phone}
          onChange={handlePhoneChange}
          placeholder="+380 (93) 321 23"
          onBlur={validationHandler}
        />
        {!validationErrors.phone.valid ? (
          <span className={styles.error}>{validationErrors.phone.message}</span>
        ) : null}
      </label>
      <label className={styles.label}>
        Email
        <input
          className={styles.input}
          type="email"
          name="email"
          value={formData.email}
          onChange={handleChangeInputs}
          required
          placeholder="email"
          onBlur={validationHandler}
        />
        {!validationErrors.email.valid ? (
          <span className={styles.error}>{validationErrors.email.message}</span>
        ) : null}
      </label>
      <label className={styles.label}>
        Feedback
        <textarea
          className={styles.input}
          name="feedback"
          rows="6"
          value={formData.feedback}
          onChange={handleChangeInputs}
          required
          minLength="100"
          maxLength="150"
          placeholder="feedback text"
          onBlur={validationHandler}
        ></textarea>
        {!validationErrors.feedback.valid ? (
          <span className={styles.error}>
            {validationErrors.feedback.message}
          </span>
        ) : null}
      </label>
      <label className={styles.ratingLabel}>
        <StarRating
          handleSetRating={handleSetRating}
          rating={formData.rating}
        />
        <p>{formData.rating}</p>
      </label>
      <label className={styles.label}>
        Department
        <select
          name="department"
          value={formData.department}
          onChange={handleChangeInputs}
          className={styles.select}
        >
          <option value="value1">Sales</option>
          <option value="value2">Delivery</option>
          <option value="value3">Support</option>
        </select>
      </label>
      <div className={styles.wrapper}>
        <label htmlFor="permission">
          - permission to process personal data
        </label>
        <input
          id="permission"
          type="checkbox"
          name="permission"
          checked={formData.permission}
          onChange={handleChangeInputs}
          required
        />
      </div>
      <button className={styles.btn} type="submit">
        Send feedback
      </button>
    </form>
  );
}

FeedbackForm.propTypes = {
  onSubmitForm: PropTypes.func.isRequired,
};
