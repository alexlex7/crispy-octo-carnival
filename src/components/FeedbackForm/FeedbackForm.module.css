.form {
  max-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  row-gap: 15px;
  overflow-y: auto;
  scrollbar-width: none;
}
.form::-webkit-scrollbar {
  display: none;
}

.input {
  padding: 8px;
  font-family: inherit;
  width: 100%;
  border: 1px solid #cacaca;
  border-radius: 5px;
  outline: none;
}

.input:focus {
  border: 1px solid #1e02bc;
}

.phoneInput:focus {
  border: 1px solid #1e02bc;
}
.phoneInput:focus + .dropdownBtn {
  border: 1px solid #1e02bc;
}
.label {
  position: relative;
  display: flex;
  flex-direction: column;
  margin-bottom: 8px;
  width: 100%;
}

.select {
  font-family: inherit;
  border: 1px solid #cacaca;
  outline: none;
  align-self: flex-start;
  border-radius: 5px;
}

.btn {
  display: inline-flex;
  margin: 0 4px;
  padding: 12px 24px;
  border: none;
  border-radius: 4px;

  font-family: sans-serif;
  font-size: 16px;

  background-color: #2196f3;
  color: #fff;
  cursor: pointer;
}

.btn:hover,
.btn:focus {
  background-color: #1976d2;
}

.error {
  position: absolute;
  bottom: -20px;
  left: 20px;
  font-size: 14px;
  color: red;
}
.wrapper {
  display: flex;
  flex-direction: row-reverse;
  justify-content: center;
  align-items: center;
  column-gap: 5px;
}

.ratingLabel {
  position: relative;
  display: flex;
  align-items: flex-end;
  column-gap: 15px;
}

.ratingLabel > p {
  position: absolute;
  right: -35px;
  top: 12px;
  font-size: larger;
  font-weight: 500;
}

@media screen and (max-width: 1204px) {
  .label {
    width: 100%;
  }
}
