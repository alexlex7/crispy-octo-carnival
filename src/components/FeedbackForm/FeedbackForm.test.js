import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import FeedbackForm from './FeedbackForm';
import userEvent from '@testing-library/user-event';

test('rendering and submitting form', async () => {
  const handleSubmit = jest.fn();
  render(<FeedbackForm onSubmitForm={handleSubmit} />);
  const feedback = userEvent.setup();

  expect(
    screen.getByRole('textbox', { name: /first name/i })
  ).toBeInTheDocument();

  expect(
    screen.getByRole('textbox', { name: /last name/i })
  ).toBeInTheDocument();

  expect(screen.getByRole('textbox', { name: /phone/i })).toBeInTheDocument();

  expect(screen.getByRole('textbox', { name: /email/i })).toBeInTheDocument();

  expect(
    screen.getByRole('textbox', { name: /feedback/i })
  ).toBeInTheDocument();

  expect(
    screen.getByRole('combobox', { name: /department/i })
  ).toBeInTheDocument();

  expect(
    screen.getByRole('checkbox', {
      name: /\- permission to process personal data/i,
    })
  ).toBeInTheDocument();

  await feedback.type(
    screen.getByRole('textbox', { name: /first name/i }),
    'Alex'
  );
  await feedback.type(
    screen.getByRole('textbox', { name: /last name/i }),
    'Dee'
  );
  await feedback.type(
    screen.getByRole('textbox', { name: /phone/i }),
    '+380 (93) 321 23'
  );
  await feedback.type(
    screen.getByRole('textbox', { name: /email/i }),
    'mail@gmail.com'
  );
  await feedback.type(
    screen.getByRole('textbox', { name: /feedback/i }),
    'Lorem ipsum dolor sit amet consectetur adipisicing elit. Rem dignissimos recusandae deserunt accusamus est illum nesciunt soluta quae iusto.'
  );
  await feedback.selectOptions(
    screen.getByRole('combobox', { name: /department/i }),
    'value2'
  );
  await feedback.click(
    screen.getByRole('checkbox', {
      name: /\- permission to process personal data/i,
    })
  );

  await feedback.click(screen.getByRole('button', { name: /Send feedback/i }));

  await waitFor(() => {
    expect(handleSubmit).toHaveBeenCalledWith({
      firstName: 'Alex',
      lastName: 'Dee',
      phone: '3809332123',
      email: 'mail@gmail.com',
      feedback:
        'Lorem ipsum dolor sit amet consectetur adipisicing elit. Rem dignissimos recusandae deserunt accusamus est illum nesciunt soluta quae iusto.',
      department: 'value2',
      permission: true,
      rating: 0,
    });
  });
});
