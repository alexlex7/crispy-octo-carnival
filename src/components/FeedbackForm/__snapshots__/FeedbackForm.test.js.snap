// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render feedback form component 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <label
      class="label"
    >
      First name
      <input
        class="input"
        minlength="1"
        name="firstName"
        placeholder="enter your first name"
        required=""
        type="text"
        value=""
      />
    </label>
    <label
      class="label"
    >
      Last name
      <input
        class="input"
        minlength="2"
        name="lastName"
        placeholder="enter your last name"
        required=""
        type="text"
        value=""
      />
    </label>
    <label
      class="label"
    >
      Phone
      <div
        class=" react-tel-input "
      >
        <div
          class="special-label"
        >
          Phone
        </div>
        <input
          class="form-control phoneInput"
          name="phone"
          placeholder="+380 (93) 321 23"
          style="width: 100%;"
          type="tel"
          value=""
        />
        <div
          class="flag-dropdown dropdownBtn"
        >
          <div
            aria-haspopup="listbox"
            class="selected-flag"
            role="button"
            tabindex="0"
            title=""
          >
            <div
              class="flag 0"
            >
              <div
                class="arrow"
              />
            </div>
          </div>
        </div>
      </div>
    </label>
    <label
      class="label"
    >
      Email
      <input
        class="input"
        name="email"
        placeholder="email"
        required=""
        type="email"
        value=""
      />
    </label>
    <label
      class="label"
    >
      Feedback
      <textarea
        class="input"
        maxlength="150"
        minlength="100"
        name="feedback"
        placeholder="feedback text"
        required=""
        rows="6"
      />
    </label>
    <label
      class="ratingLabel"
    >
      <div
        class="ratingWrapper"
      >
        <div
          class="starsWrapper"
        >
          <svg
            class="icon"
            fill="none"
          >
            star.svg
          </svg>
          <svg
            class="icon"
            fill="none"
          >
            star.svg
          </svg>
          <svg
            class="icon"
            fill="none"
          >
            star.svg
          </svg>
          <svg
            class="icon"
            fill="none"
          >
            star.svg
          </svg>
          <svg
            class="icon"
            fill="none"
          >
            star.svg
          </svg>
        </div>
        <div
          class="starsActiveWrapper"
          style="width: 0%;"
        >
          <div
            style="display: flex;"
          >
            <svg
              class="activeIcon"
              fill="orange"
            >
              star.svg
            </svg>
            <svg
              class="activeIcon"
              fill="orange"
            >
              star.svg
            </svg>
            <svg
              class="activeIcon"
              fill="orange"
            >
              star.svg
            </svg>
            <svg
              class="activeIcon"
              fill="orange"
            >
              star.svg
            </svg>
            <svg
              class="activeIcon"
              fill="orange"
            >
              star.svg
            </svg>
          </div>
        </div>
      </div>
      <p>
        0
      </p>
    </label>
    <label
      class="label"
    >
      Department
      <select
        class="select"
        name="department"
      >
        <option
          value="value1"
        >
          Sales
        </option>
        <option
          value="value2"
        >
          Delivery
        </option>
        <option
          value="value3"
        >
          Support
        </option>
      </select>
    </label>
    <div
      class="wrapper"
    >
      <label>
        - permission to process personal data
      </label>
      <input
        name="permission"
        required=""
        type="checkbox"
      />
    </div>
    <button
      class="btn"
      type="submit"
    >
      Send feedback
    </button>
  </form>
</DocumentFragment>
`;
